***Подготовка данных для загрузки в модель AW.  Преобразование, очистка, конвертация, группировка данных, конструирование признаков.***

In [1]:
# импорт библиотек для анализа и обработки данных
import pandas as pd
import numpy as np

pd.options.mode.chained_assignment = None

**DONORS**

In [2]:
# загружаем 1-й файл датасета
df_donors  = pd.read_csv('Доноры.csv',sep=';')

In [3]:
df_donors.head()

,id,sex
0,000046e6a3cca84517b4708f86a30eda,NaN
1,00006d1acf5e723c1437c32d65a3b754,NaN
2,000085233064aac73f54ed9292e76613,f
3,000085233064aac73f54ed9292e76613,f
4,0000e5a0eec69bfb14012534891b8cb9,NaN


In [4]:
df_donors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234662 entries, 0 to 234661
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   id      234662 non-null  object
 1   sex     73233 non-null   object
dtypes: object(2)
memory usage: 3.6+ MB


In [5]:
df_donors.describe()

,id,sex
count,234662,73233
unique,232076,2
top,1a5462448f458af76fcfea87badfaafb,f
freq,15,53939


в таблице "Доноры" пол заявлен только у примерно 31% записей, при этом подавляющее число ID доноров - уникальны (99%). 

In [6]:
# удаляем дубликаты id доноров, оставляя только уникальные значения
df_donors = df_donors.drop_duplicates(['id'])

In [7]:
df_donors.describe()
# после удаления дубликатов картина не меняется. Только у 31% доноров заявлен пол. 


,id,sex
count,232076,72126
unique,232076,2
top,000046e6a3cca84517b4708f86a30eda,f
freq,1,53180


**DONATIONS**

In [8]:
# загружаем 1-й файл датасета
df_donations = pd.read_csv('Пожертвования.csv', sep=';', engine='python')

In [9]:
df_donations.head()

,id,donor_id,purpose_id,transaction_id,recurring_id,status,when,amount,amount_total,category,device_type,gateway,os
0,60e72b5a985f702c62fef16ea95d403d,70a16408d291cbe843893174919a561a,00668da16a708df9a54fa76ef97d5e77,79c76370621c03d9dee3cc57379ffa8d,NaN,3,2012-02-25 17:05:00,1000,1000,BankCard,NaN,robokassa,NaN
1,802a10be56e88026eaa9aa773d0890c8,7c9bc4fb1bb133cef5a68710fff911d8,00668da16a708df9a54fa76ef97d5e77,5d6784d7a62353300844fc5e3d8b8764,NaN,3,2012-02-25 19:14:00,400,400,BankCard,NaN,robokassa,NaN
2,0dfd96e197c13c27b462f2d95342a757,6281148700f119139418212037fe85aa,c442f6d4d05db5941a5642e8fbabc7b1,0ca4d2acd9c7c00d483b58889c286454,NaN,3,2012-02-27 11:23:00,1,1,BankCard,NaN,robokassa,NaN
3,e4c978174bab088581c1731115d212f9,faee8f97a290186dcb6a4c798550637f,c442f6d4d05db5941a5642e8fbabc7b1,4baf7330b63ec9f5e7aa41c61719469d,NaN,3,2012-03-01 13:56:00,200,200,BANKOCEAN2,NaN,robokassa,NaN
4,8fc5984ba863dcc83d32a30f326d4821,406d9953fd5ed6c1ca968d867b68b1af,c442f6d4d05db5941a5642e8fbabc7b1,db5ec80608fd302967c374f5e143c9dc,NaN,3,2012-03-01 19:06:00,1000,1000,BANKOCEAN2,NaN,robokassa,NaN


In [10]:
df_donations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 472987 entries, 0 to 472986
Data columns (total 13 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   id              472987 non-null  object
 1   donor_id        472987 non-null  object
 2   purpose_id      472896 non-null  object
 3   transaction_id  472987 non-null  object
 4   recurring_id    360115 non-null  object
 5   status          472987 non-null  int64 
 6   when            472987 non-null  object
 7   amount          472967 non-null  object
 8   amount_total    472987 non-null  object
 9   category        471729 non-null  object
 10  device_type     17414 non-null   object
 11  gateway         472987 non-null  object
 12  os              17414 non-null   object
dtypes: int64(1), object(12)
memory usage: 46.9+ MB


In [11]:
df_donations.describe()

,status
count,472987.000000
mean,3.430158
std,0.964960
min,-1.000000
25%,3.000000
50%,3.000000
75%,3.000000
max,6.000000


**TRANSACTIONS**

In [12]:
# загружаем 3-й файл датасета
df_transactions = pd.read_csv('Транзакции.csv', sep=';', engine='python')

In [13]:
df_transactions.head()

,id,when,amount,currency,ipcountry,ipregion
0,090b0493789e471517272a2608a7d752,2012-10-29 16:29:32,1,RUB,RU,Москва
1,dfead3e439ed0a2d06d57d5267506afa,2012-10-29 16:54:10,10000,RUB,RU,Москва
2,90b5b89f52513f9884e51992cf52a280,2012-10-29 17:29:18,1000,RUB,RU,Московская область
3,118d2ad46f4cd7aacb33d4d4c3245519,2012-10-29 17:32:36,100,RUB,RU,Саратовская область
4,f5a5940b3d4f0ffc6b3cbc02c17d93f8,2012-10-29 17:36:52,500,RUB,RU,Московская область


In [14]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444556 entries, 0 to 444555
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   id         444556 non-null  object
 1   when       444556 non-null  object
 2   amount     444556 non-null  object
 3   currency   444556 non-null  object
 4   ipcountry  350690 non-null  object
 5   ipregion   324451 non-null  object
dtypes: object(6)
memory usage: 20.4+ MB


In [15]:
df_transactions.describe()

,id,when,amount,currency,ipcountry,ipregion
count,444556,444556,444556,444556,350690,324451
unique,444556,295923,1642,2,170,886
top,090b0493789e471517272a2608a7d752,2019-10-10 10:09:51,100,RUB,RU,Москва
freq,1,212,89068,408780,318697,96616


**PURPOSES**

In [16]:
# загружаем 4-й файл датасета
df_purposes = pd.read_csv('Цели сборов.csv',sep=';')

In [17]:
df_purposes.head()

,id,fund_id
0,00668da16a708df9a54fa76ef97d5e77,e3fd213a01aee164f475152137a83eec
1,00991be67ae4c521edd6deca31d46c70,c06d416f1b315dc3c09b281168df1a4f
2,01ca25e4c3fa621c422079b5c1e74819,0358fd748122e743e8e54305ac933c29
3,01f3dbd827188bf292c83f81ac775444,c06d416f1b315dc3c09b281168df1a4f
4,01f9234d553e596e83fce706e3efae39,e3fd213a01aee164f475152137a83eec


In [18]:
df_purposes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       1025 non-null   object
 1   fund_id  1025 non-null   object
dtypes: object(2)
memory usage: 16.1+ KB


In [19]:
df_purposes.describe()

,id,fund_id
count,1025,1025
unique,1025,35
top,00668da16a708df9a54fa76ef97d5e77,e3fd213a01aee164f475152137a83eec
freq,1,455


**MERGE - объединение таблиц**

In [20]:
# начинаем объединять загруженные таблицы
df_merged = pd.merge(df_donations, df_donors, how='left', left_on='donor_id', right_on='id')

In [21]:
# смотрим корректность объединения
df_merged

,id_x,donor_id,purpose_id,transaction_id,recurring_id,status,when,amount,amount_total,category,device_type,gateway,os,id_y,sex
0,60e72b5a985f702c62fef16ea95d403d,70a16408d291cbe843893174919a561a,00668da16a708df9a54fa76ef97d5e77,79c76370621c03d9dee3cc57379ffa8d,NaN,3,2012-02-25 17:05:00,1000,1000,BankCard,NaN,robokassa,NaN,70a16408d291cbe843893174919a561a,m
1,802a10be56e88026eaa9aa773d0890c8,7c9bc4fb1bb133cef5a68710fff911d8,00668da16a708df9a54fa76ef97d5e77,5d6784d7a62353300844fc5e3d8b8764,NaN,3,2012-02-25 19:14:00,400,400,BankCard,NaN,robokassa,NaN,7c9bc4fb1bb133cef5a68710fff911d8,f
2,0dfd96e197c13c27b462f2d95342a757,6281148700f119139418212037fe85aa,c442f6d4d05db5941a5642e8fbabc7b1,0ca4d2acd9c7c00d483b58889c286454,NaN,3,2012-02-27 11:23:00,1,1,BankCard,NaN,robokassa,NaN,6281148700f119139418212037fe85aa,m
3,e4c978174bab088581c1731115d212f9,faee8f97a290186dcb6a4c798550637f,c442f6d4d05db5941a5642e8fbabc7b1,4baf7330b63ec9f5e7aa41c61719469d,NaN,3,2012-03-01 13:56:00,200,200,BANKOCEAN2,NaN,robokassa,NaN,faee8f97a290186dcb6a4c798550637f,m
4,8fc5984ba863dcc83d32a30f326d4821,406d9953fd5ed6c1ca968d867b68b1af,c442f6d4d05db5941a5642e8fbabc7b1,db5ec80608fd302967c374f5e143c9dc,NaN,3,2012-03-01 19:06:00,1000,1000,BANKOCEAN2,NaN,robokassa,NaN,406d9953fd5ed6c1ca968d867b68b1af,f
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
472982,71e9fe7a34fd8cf0e0127ab878cbd217,9e92240590475fad9e95f52ecbcb128b,b595e4581542224fa6eaedf161ace910,95d0c24e384b358c9f27e75a8eb2a994,63e8f4e9c756353a28396d59604e50b1,3,2021-06-30 23:30:00,487,500,card,NaN,cloudpayments,NaN,NaN,NaN
472983,cd470f9538a084046cfdbbbcd37faa34,c5e9399066e795386f69b4c0a0e9ccf6,b0cecf1a7fecaad80a37bac7198d9515,afe1ab70e3edca3da162866f9331953d,832a276868dddd624c5842c2ff4ab39f,5,2021-06-30 23:31:21,200,200,card,NaN,cloudpayments,NaN,c5e9399066e795386f69b4c0a0e9ccf6,f
472984,cad122bff720e5dd2f0fb380077ff9b1,1d73d99fcb783d2b66d2aceb6afe2377,a95a01351be774d85a7cd6178f4adab5,0f2d82808182a420b34c098b80a7eac3,NaN,3,2021-06-30 23:35:00,2883,3000,card,NaN,cloudpayments,NaN,NaN,NaN
472985,fc88c6bdcfed4fa2808f4b39f5577b9d,945004b40a0e609b803265ac63d8e791,ffe1cab8488e89ddbc85a09b62c3070e,3afa15b6b390ce66fc4be6152f90356b,NaN,-1,2021-06-30 23:47:19,"106,92",110,card,mobile,mixplat,iOS,945004b40a0e609b803265ac63d8e791,f


In [22]:
# удяляем дублирующийся столбец
df_merged = df_merged.drop('id_y', axis=1)

In [23]:
df_merged['donor_id'].describe()

count                               472987
unique                              211393
top       66b7587d1062a1200e41ce02e8582b60
freq                                  2380
Name: donor_id, dtype: object

In [24]:
# Выполняем следующее объединение таблиц. На этот раз дублирующийся столбец с id удаляем сразу
df_merged = pd.merge(df_merged, df_purposes, how='left', left_on='purpose_id', right_on='id').drop('id', axis=1)

In [25]:
df_merged['fund_id'].describe()

count                               472896
unique                                  19
top       497b5be5c4c5bf0e9a6c3791c01476f5
freq                                256992
Name: fund_id, dtype: object

In [26]:
# Выполняем последнее объединение имеющихся в датасете таблиц
df_merged = pd.merge(df_merged, df_transactions, how='left', left_on='transaction_id', right_on='id').drop('id', axis=1)

In [27]:
df_merged.head()
# обогатили основную таблицу (Пожертвования) данными из остальных таблиц. Получили 1 объединенный датасет сырых данных.

,id_x,donor_id,purpose_id,transaction_id,recurring_id,status,when_x,amount_x,amount_total,category,device_type,gateway,os,sex,fund_id,when_y,amount_y,currency,ipcountry,ipregion
0,60e72b5a985f702c62fef16ea95d403d,70a16408d291cbe843893174919a561a,00668da16a708df9a54fa76ef97d5e77,79c76370621c03d9dee3cc57379ffa8d,NaN,3,2012-02-25 17:05:00,1000,1000,BankCard,NaN,robokassa,NaN,m,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN,NaN,NaN
1,802a10be56e88026eaa9aa773d0890c8,7c9bc4fb1bb133cef5a68710fff911d8,00668da16a708df9a54fa76ef97d5e77,5d6784d7a62353300844fc5e3d8b8764,NaN,3,2012-02-25 19:14:00,400,400,BankCard,NaN,robokassa,NaN,f,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN,NaN,NaN
2,0dfd96e197c13c27b462f2d95342a757,6281148700f119139418212037fe85aa,c442f6d4d05db5941a5642e8fbabc7b1,0ca4d2acd9c7c00d483b58889c286454,NaN,3,2012-02-27 11:23:00,1,1,BankCard,NaN,robokassa,NaN,m,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN,NaN,NaN
3,e4c978174bab088581c1731115d212f9,faee8f97a290186dcb6a4c798550637f,c442f6d4d05db5941a5642e8fbabc7b1,4baf7330b63ec9f5e7aa41c61719469d,NaN,3,2012-03-01 13:56:00,200,200,BANKOCEAN2,NaN,robokassa,NaN,m,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN,NaN,NaN
4,8fc5984ba863dcc83d32a30f326d4821,406d9953fd5ed6c1ca968d867b68b1af,c442f6d4d05db5941a5642e8fbabc7b1,db5ec80608fd302967c374f5e143c9dc,NaN,3,2012-03-01 19:06:00,1000,1000,BANKOCEAN2,NaN,robokassa,NaN,f,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN,NaN,NaN


In [28]:
# Сводная информация по объединенному датасету
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 472987 entries, 0 to 472986
Data columns (total 20 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   id_x            472987 non-null  object
 1   donor_id        472987 non-null  object
 2   purpose_id      472896 non-null  object
 3   transaction_id  472987 non-null  object
 4   recurring_id    360115 non-null  object
 5   status          472987 non-null  int64 
 6   when_x          472987 non-null  object
 7   amount_x        472967 non-null  object
 8   amount_total    472987 non-null  object
 9   category        471729 non-null  object
 10  device_type     17414 non-null   object
 11  gateway         472987 non-null  object
 12  os              17414 non-null   object
 13  sex             163870 non-null  object
 14  fund_id         472896 non-null  object
 15  when_y          444414 non-null  object
 16  amount_y        444414 non-null  object
 17  currency        444414 non-nu

**Очистка данных**

Удалим поля, дублирующие информацию в датасете.
Информация в when_y дублирует when_x. 
Amount_y дублирует amount_total.
Amount_x так же удалим т.к. величиной комиссии, во-первых, в большинстве случае можно принебречь, а во вторых она все равно входит в фактически понесенные донором расходы на пожертвование.


In [29]:
# Удалям поля, дублирующие информацию в датасете. Создаем при этом новый "очищенный" датафрейм.
df_cleaned = df_merged.drop(['when_y', 'amount_y', 'amount_x'], axis=1)

In [30]:
df_cleaned.head()

,id_x,donor_id,purpose_id,transaction_id,recurring_id,status,when_x,amount_total,category,device_type,gateway,os,sex,fund_id,currency,ipcountry,ipregion
0,60e72b5a985f702c62fef16ea95d403d,70a16408d291cbe843893174919a561a,00668da16a708df9a54fa76ef97d5e77,79c76370621c03d9dee3cc57379ffa8d,NaN,3,2012-02-25 17:05:00,1000,BankCard,NaN,robokassa,NaN,m,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN
1,802a10be56e88026eaa9aa773d0890c8,7c9bc4fb1bb133cef5a68710fff911d8,00668da16a708df9a54fa76ef97d5e77,5d6784d7a62353300844fc5e3d8b8764,NaN,3,2012-02-25 19:14:00,400,BankCard,NaN,robokassa,NaN,f,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN
2,0dfd96e197c13c27b462f2d95342a757,6281148700f119139418212037fe85aa,c442f6d4d05db5941a5642e8fbabc7b1,0ca4d2acd9c7c00d483b58889c286454,NaN,3,2012-02-27 11:23:00,1,BankCard,NaN,robokassa,NaN,m,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN
3,e4c978174bab088581c1731115d212f9,faee8f97a290186dcb6a4c798550637f,c442f6d4d05db5941a5642e8fbabc7b1,4baf7330b63ec9f5e7aa41c61719469d,NaN,3,2012-03-01 13:56:00,200,BANKOCEAN2,NaN,robokassa,NaN,m,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN
4,8fc5984ba863dcc83d32a30f326d4821,406d9953fd5ed6c1ca968d867b68b1af,c442f6d4d05db5941a5642e8fbabc7b1,db5ec80608fd302967c374f5e143c9dc,NaN,3,2012-03-01 19:06:00,1000,BANKOCEAN2,NaN,robokassa,NaN,f,e3fd213a01aee164f475152137a83eec,NaN,NaN,NaN


Далее смотрим на поля 'device_type' и 'os' они имеют не NULL значения только примерно в 4% строк. Это очень мало. При этом для обоих полей мы имеем
одинаковое количество не NULL записей (17414). Скорее всего это одни и те же 17414 строк. Удалим и эти поля.

In [31]:
df_cleaned['os'].value_counts()

os
Android    9227
iOS        5363
Windows    2493
Mac         212
Linux       119
Name: count, dtype: int64

In [32]:
df_cleaned['os'].count()

17414

In [33]:
df_cleaned['device_type'].value_counts()

device_type
mobile     14579
desktop     2790
tablet        45
Name: count, dtype: int64

In [34]:
df_cleaned['device_type'].count()

17414

In [35]:
df_cleaned = df_cleaned.drop(['device_type', 'os'], axis=1)

Далее смотрим на поле currency. Видим что в нем около 94% не NULL записей, которые принимают только 2 значения 'rub' и 'RUB' т.е. фактически одно: Рубли.
Из чего делаем вывод, что все 94% имеющихся записей одинаковы, а оставшиеся 6% с большой долей вероятности тоже являются рублями. Других варинатов у нас попросту нет. Поэтому будем считать, что все все записи в поле currency имеют единственно значение: рубли. Соответственно этот признак мы так же исключим из нашей будущей модели машинного обучения.<br> Поле id_x (id пожертвования) имеет уникальное значение для каждой записи. Соответствеено, оно так же не имеет ценности для нашей модели и будет удалено.

In [36]:
df_cleaned['currency'].value_counts()

currency
RUB    408638
rub     35776
Name: count, dtype: int64

In [37]:
df_cleaned = df_cleaned.drop(['currency', 'id_x'], axis=1)

In [38]:
# смотрим состав данных в поле 'ipcountry'
df_cleaned['ipcountry'].value_counts()

ipcountry
RU          318617
Россия       22615
US            1228
DE             831
FR             611
             ...  
TM               1
Гонконг          1
IQ               1
Мексика          1
Словения         1
Name: count, Length: 170, dtype: int64

In [39]:
# смотрим состав данных в поле 'ipregion'
df_cleaned['ipregion'].value_counts()

ipregion
Москва                    96582
Санкт-Петербург           26914
Краснодарский край        14187
Московская обл.            7593
Новосибирская обл.         5762
                          ...  
Buenos Aires F.D.             1
Provincia de Pichincha        1
Кировоградская область        1
Kocaeli                       1
Бишкек Шаары                  1
Name: count, Length: 886, dtype: int64

In [40]:
# смотрим состав данных в поле 'transaction_id'
df_cleaned['transaction_id'].value_counts()

transaction_id
79c76370621c03d9dee3cc57379ffa8d    1
6522e5ca74c1d21cb384e4c0e358e782    1
83fa687ecdab784c2c96059586113a14    1
dd5cdb92433f78927fa7bd7bc2422624    1
6210c308016d15a35bff796a28bb8051    1
                                   ..
194ef81a1fc64736d241f750571c01ad    1
df4a2730272a8b4bca54ada98dc3b1d9    1
1f8f4d064bc7c6201ca334b2e3ea5d3d    1
764b3a534b67a69ba33beeb5a27451ca    1
44391540d57972fd1cc01099aec0af54    1
Name: count, Length: 472987, dtype: int64

поле transaction_id так же хранит уникальные значения для каждой записи, поэтому его так же удаляем из датафрейма

In [41]:
df_cleaned = df_cleaned.drop('transaction_id', axis=1)

**Преобразование типов данных** 

Выше мы уже видели, что многие поля имеют у нас неподходящие для них типы данных. <br>
Исправим же это!

In [42]:
# создаем новую копию датафрейма с именем df_converted
df_converted = df_cleaned.copy()

In [43]:
# конвертируем поле с датой пожертвования из текста в datetime
df_converted['when_x'] = pd.to_datetime(df_converted['when_x'])

In [44]:
df_converted.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 472987 entries, 0 to 472986
Data columns (total 12 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   donor_id      472987 non-null  object        
 1   purpose_id    472896 non-null  object        
 2   recurring_id  360115 non-null  object        
 3   status        472987 non-null  int64         
 4   when_x        472987 non-null  datetime64[ns]
 5   amount_total  472987 non-null  object        
 6   category      471729 non-null  object        
 7   gateway       472987 non-null  object        
 8   sex           163870 non-null  object        
 9   fund_id       472896 non-null  object        
 10  ipcountry     350551 non-null  object        
 11  ipregion      324338 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(10)
memory usage: 43.3+ MB


In [45]:
# заменяем в значениях колонки 'amount_total' запятые на точки перед преобразованием в int
df_converted['amount_total'] =  df_converted['amount_total'].str.replace(',', '.')

In [46]:
# преобразовавываем в Int
df_converted['amount_total'] = df_converted['amount_total'].astype(float)
df_converted['amount_total'] = df_converted['amount_total'].astype(int)

In [47]:
# добавляем поле is_reccurent, в котором указываем является ли пожертвование реккурентынм или нет.
# Оно является таковым, если у него имеется recurring_id.
df_converted['is_reccurent'] = np.where(df_converted['recurring_id'].notna() & df_converted['recurring_id'].str.strip().astype(bool), 1, 0)

In [48]:
# смотрим ТОП количества транзакций у доноров.
df_converted['donor_id'].value_counts()

donor_id
66b7587d1062a1200e41ce02e8582b60    2380
42dd83ab04d0c7fcf050faa31adb3074    1315
9d33c44b55783f799e277eeb5856e8b7     854
8e2739b8906f749234e4a3d82344b5ee     806
91ec49a85540354266461f585ab6167e     778
                                    ... 
cb406baa6a152a24a9a46d21b3d60768       1
d46a6ca05a3f5b63a55828c4c68d1309       1
0b8e8d1547c78a2a4a7f6e73878e6a0b       1
e5a764794951084a1ef1885d19970a37       1
8fe2b0c0e7f89ee2d5f10f3ff9b8d8a8       1
Name: count, Length: 211393, dtype: int64

**Обрабатываем значения NULL (NaN)**

т.к. значения NULL у нас только в полях с текстовыми данными, то заменим их на 'unknown'. Поиск средних и каких-то наиболее подходящих значений не видится возможным в данном случае

In [49]:

df_final= df_converted.fillna('unknown')

In [50]:
# убеждаемся, что NULL значений больше нет
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 472987 entries, 0 to 472986
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   donor_id      472987 non-null  object        
 1   purpose_id    472987 non-null  object        
 2   recurring_id  472987 non-null  object        
 3   status        472987 non-null  int64         
 4   when_x        472987 non-null  datetime64[ns]
 5   amount_total  472987 non-null  int32         
 6   category      472987 non-null  object        
 7   gateway       472987 non-null  object        
 8   sex           472987 non-null  object        
 9   fund_id       472987 non-null  object        
 10  ipcountry     472987 non-null  object        
 11  ipregion      472987 non-null  object        
 12  is_reccurent  472987 non-null  int32         
dtypes: datetime64[ns](1), int32(2), int64(1), object(9)
memory usage: 43.3+ MB


In [51]:
df_final.describe()

,status,when_x,amount_total,is_reccurent
count,472987.000000,472987,472987.000000,472987.000000
mean,3.430158,2019-10-03 04:11:53.739872768,634.263878,0.761363
min,-1.000000,2012-02-25 17:05:00,-487500.000000,0.000000
25%,3.000000,2019-02-02 20:26:00,100.000000,1.000000
50%,3.000000,2020-07-08 11:28:52,300.000000,1.000000
75%,3.000000,2021-02-09 08:02:25.500000,500.000000,1.000000
max,6.000000,2021-06-30 23:49:27,858150.000000,1.000000
std,0.964960,NaN,3893.577715,0.426251


Получаем в итоге 3 варианта пола донора. М, Ж и 'пол неизвестен'. Этот столбец решено не убирать из признаков т.к. это фактически единственная персональная информация о самом доноре ( о малой части доноров), которая имеется у нас.

In [52]:
df_final['sex'].value_counts()

sex
unknown    309117
f          118105
m           45765
Name: count, dtype: int64

In [53]:
# Обратим внимание на колонку 'recurring_id'
df_final['recurring_id'].value_counts()

recurring_id
d41d8cd98f00b204e9800998ecf8427e    236110
unknown                             112872
8b231a6be88fae53962ab7985432cf1d       806
54884e5b8c8dfa7b1ad9481524f2c627       747
9c905acb96dfc1e62c77fb9c2790e917       746
                                     ...  
12e5a1e693337df1535386e74c84e84b         1
6ac2a6606e5c543012493a46d79a1f36         1
59556ab8dc546f60a4fcb9bb656c1904         1
e794b72e27d6c5c6ffa8a757b3d22aab         1
3ec1947c2da80ebc2de2850a19fba2f7         1
Name: count, Length: 14894, dtype: int64

In [54]:
# Для колонки 'recurring_id' заменим 'unknown' на 'no' т.к. id фактически отсутствуют по причине нереккурентности операции, а не являются неизвестными.
df_final['recurring_id'] = df_final['recurring_id'].str.replace('unknown', 'no')

In [55]:
# Смотрим окончательный вариант датафрейма с сырыми очищенными данными
df_final

,donor_id,purpose_id,recurring_id,status,when_x,amount_total,category,gateway,sex,fund_id,ipcountry,ipregion,is_reccurent
0,70a16408d291cbe843893174919a561a,00668da16a708df9a54fa76ef97d5e77,no,3,2012-02-25 17:05:00,1000,BankCard,robokassa,m,e3fd213a01aee164f475152137a83eec,unknown,unknown,0
1,7c9bc4fb1bb133cef5a68710fff911d8,00668da16a708df9a54fa76ef97d5e77,no,3,2012-02-25 19:14:00,400,BankCard,robokassa,f,e3fd213a01aee164f475152137a83eec,unknown,unknown,0
2,6281148700f119139418212037fe85aa,c442f6d4d05db5941a5642e8fbabc7b1,no,3,2012-02-27 11:23:00,1,BankCard,robokassa,m,e3fd213a01aee164f475152137a83eec,unknown,unknown,0
3,faee8f97a290186dcb6a4c798550637f,c442f6d4d05db5941a5642e8fbabc7b1,no,3,2012-03-01 13:56:00,200,BANKOCEAN2,robokassa,m,e3fd213a01aee164f475152137a83eec,unknown,unknown,0
4,406d9953fd5ed6c1ca968d867b68b1af,c442f6d4d05db5941a5642e8fbabc7b1,no,3,2012-03-01 19:06:00,1000,BANKOCEAN2,robokassa,f,e3fd213a01aee164f475152137a83eec,unknown,unknown,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
472982,9e92240590475fad9e95f52ecbcb128b,b595e4581542224fa6eaedf161ace910,63e8f4e9c756353a28396d59604e50b1,3,2021-06-30 23:30:00,500,card,cloudpayments,unknown,93d54d49e8d73a64490700aafe8909c4,unknown,unknown,1
472983,c5e9399066e795386f69b4c0a0e9ccf6,b0cecf1a7fecaad80a37bac7198d9515,832a276868dddd624c5842c2ff4ab39f,5,2021-06-30 23:31:21,200,card,cloudpayments,f,e3fd213a01aee164f475152137a83eec,unknown,unknown,1
472984,1d73d99fcb783d2b66d2aceb6afe2377,a95a01351be774d85a7cd6178f4adab5,no,3,2021-06-30 23:35:00,3000,card,cloudpayments,unknown,9161f2362b847ca9307ca394431ead09,RU,unknown,0
472985,945004b40a0e609b803265ac63d8e791,ffe1cab8488e89ddbc85a09b62c3070e,no,-1,2021-06-30 23:47:19,110,card,mixplat,f,e3fd213a01aee164f475152137a83eec,unknown,unknown,0


**Группировка по donor_id**

Далее будем группировать данные по donor_id, ведь наша модель машинного обучения будет предсказывать вероятность ухода в отток для каждого донора нащего датасета. Соответственно каждая запись нашего датасета на входе модели машинного обучения должна соответствовать уникальному donor_id.

In [56]:
df_final['donor_id'].value_counts()

donor_id
66b7587d1062a1200e41ce02e8582b60    2380
42dd83ab04d0c7fcf050faa31adb3074    1315
9d33c44b55783f799e277eeb5856e8b7     854
8e2739b8906f749234e4a3d82344b5ee     806
91ec49a85540354266461f585ab6167e     778
                                    ... 
cb406baa6a152a24a9a46d21b3d60768       1
d46a6ca05a3f5b63a55828c4c68d1309       1
0b8e8d1547c78a2a4a7f6e73878e6a0b       1
e5a764794951084a1ef1885d19970a37       1
8fe2b0c0e7f89ee2d5f10f3ff9b8d8a8       1
Name: count, Length: 211393, dtype: int64

перед группировкой по donor_id оставим  в датафрейме только транзакции, имеющие статусы '3' и '5', т.е. завершенные либо отмененные т.к. они составляют почти 99% вcех транзакций. (Статус '-1' (около 1% транзакций) и статус '6' вообще не имеют описания, поэтому их удаляем из датасета. Статусы '1', '2' являются промежуточными и имеют менее 0.25%, статус 4 встречается менее чем в 0.1% случаев. Удаляем и их.)

In [57]:
#Создадим для группировки новый отдельный датафрейм
df_grouped = df_final[(df_final['status'] == 3) | (df_final['status'] == 5)]

In [58]:
# выделяем дату без времени из колонки 'when_x'
df_grouped['date'] = df_grouped['when_x'].dt.date


In [59]:
# Удаляем колонку 'when_x'. Она нам больше не нужны
df_grouped.drop('when_x', axis=1, inplace=True)

In [60]:
# преобразуем свежесозданную колонку с датой в тип данных 'дата'
df_grouped['date'] = pd.to_datetime(df_grouped['date'])

In [61]:
# проверим соответствие типов данных
df_grouped.info()

<class 'pandas.core.frame.DataFrame'>
Index: 466828 entries, 0 to 472986
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   donor_id      466828 non-null  object        
 1   purpose_id    466828 non-null  object        
 2   recurring_id  466828 non-null  object        
 3   status        466828 non-null  int64         
 4   amount_total  466828 non-null  int32         
 5   category      466828 non-null  object        
 6   gateway       466828 non-null  object        
 7   sex           466828 non-null  object        
 8   fund_id       466828 non-null  object        
 9   ipcountry     466828 non-null  object        
 10  ipregion      466828 non-null  object        
 11  is_reccurent  466828 non-null  int32         
 12  date          466828 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int32(2), int64(1), object(9)
memory usage: 46.3+ MB


In [62]:
# создадим первый промежуточный датафрейм и вычислим значения полей для успешных транзакций (статус=3)
df_grouped_1 = (
    df_grouped[df_grouped['status'] == 3]
    .groupby('donor_id', as_index=False)
    .agg(
        sex=('sex', 'first'),  # Предположим, что пол одинаковый для одного donor_id
        sum_amount=('amount_total', 'sum'),  # сумма всех сделанных пожертований для каждого донора
        mean_amount=('amount_total', 'mean'),  # средний размер пожертования для каждого донора
        count_transactions_completed = ('amount_total', 'count'),   # количество успешных транзакций ( пожертвований)
        count_reccurent_transactions = ('is_reccurent', 'sum'),  # количсетво успешных реккурентных транзакций
        mean_date_diff = ('date', lambda dates: (dates.sort_values().diff().mean().days)),   # средняя разница в днях между успешными транзакциями
        first_transaction_date = ('date', 'min'),   # дата первой успешной транзакции донора
        last_transaction_date = ('date', 'max')     # дата последней успешной транзакции донора
    )
)

In [63]:
# округлим значения в колонке 'mean_amount'
df_grouped_1['mean_amount'] = round(df_grouped_1['mean_amount'], 2)

In [64]:
# создадим второй промежуточный датафрейм для подсчета количества отмененных транзакций (статус=5)
df_grouped_2 = (
    df_grouped[df_grouped['status'] == 5]
    .groupby('donor_id', as_index=False)
    .agg(
        count_transactions_cancelled = ('amount_total', 'count')
    )
)

In [65]:
# объединяем 2 датафрейма в 1. ( Добавляем колонку с количествои отклоненных платежей к первому датафрейму)
df_grouped  = pd.merge(df_grouped_1, df_grouped_2, how='left', on='donor_id')

In [66]:
df_grouped

,donor_id,sex,sum_amount,mean_amount,count_transactions_completed,count_reccurent_transactions,mean_date_diff,first_transaction_date,last_transaction_date,count_transactions_cancelled
0,000085233064aac73f54ed9292e76613,f,6330,2110.00,3,0,610.0,2017-12-18,2021-04-22,NaN
1,00008e5479b05b0cf84620b017a54f6e,unknown,200,200.00,1,1,NaN,2020-05-04,2020-05-04,NaN
2,0000f19ad7b4da36765dacad76cdadc9,f,800,400.00,2,0,35.0,2017-10-03,2017-11-07,NaN
3,0000f6c6b8f0a8c8cdbcaa2330ea7f64,f,8600,296.55,29,28,31.0,2018-12-23,2021-06-02,NaN
4,0001969a913d9ada268d363b54afa383,unknown,50,50.00,1,1,NaN,2019-12-10,2019-12-10,NaN
...,...,...,...,...,...,...,...,...,...,...
168373,fffdc2d9ebcd2b03740a6bcfe9461a5b,f,100,100.00,1,1,NaN,2021-06-14,2021-06-14,NaN
168374,fffe4b8e50a941e677ec80535eaba4f9,unknown,300,300.00,1,1,NaN,2020-03-04,2020-03-04,NaN
168375,fffeb1bf939cf19731a442e92f9a2175,f,8500,500.00,17,16,38.0,2018-12-20,2020-08-20,1.0
168376,ffff964599367cf1498f1e19aa3b0337,f,500,500.00,1,0,NaN,2021-06-18,2021-06-18,NaN


In [67]:
df_grouped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 168378 entries, 0 to 168377
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   donor_id                      168378 non-null  object        
 1   sex                           168378 non-null  object        
 2   sum_amount                    168378 non-null  int32         
 3   mean_amount                   168378 non-null  float64       
 4   count_transactions_completed  168378 non-null  int64         
 5   count_reccurent_transactions  168378 non-null  int32         
 6   mean_date_diff                34718 non-null   float64       
 7   first_transaction_date        168378 non-null  datetime64[ns]
 8   last_transaction_date         168378 non-null  datetime64[ns]
 9   count_transactions_cancelled  27110 non-null   float64       
dtypes: datetime64[ns](2), float64(3), int32(2), int64(1), object(2)
memory usage: 11

In [68]:
# заменяем значения NaN в колонке с количествои отмененных транзакций на 0.
df_grouped['count_transactions_cancelled'] = df_grouped['count_transactions_cancelled'].fillna(0).astype(int)

In [69]:
# Добавим столбец с процентом отменненных транзакций
df_grouped['cancelled_percentage'] = 100 * round(df_grouped['count_transactions_cancelled'] / (df_grouped['count_transactions_completed'] + df_grouped['count_transactions_cancelled']), 3)

In [70]:
# проверим результат
df_grouped.head()

,donor_id,sex,sum_amount,mean_amount,count_transactions_completed,count_reccurent_transactions,mean_date_diff,first_transaction_date,last_transaction_date,count_transactions_cancelled,cancelled_percentage
0,000085233064aac73f54ed9292e76613,f,6330,2110.00,3,0,610.0,2017-12-18,2021-04-22,0,0.0
1,00008e5479b05b0cf84620b017a54f6e,unknown,200,200.00,1,1,NaN,2020-05-04,2020-05-04,0,0.0
2,0000f19ad7b4da36765dacad76cdadc9,f,800,400.00,2,0,35.0,2017-10-03,2017-11-07,0,0.0
3,0000f6c6b8f0a8c8cdbcaa2330ea7f64,f,8600,296.55,29,28,31.0,2018-12-23,2021-06-02,0,0.0
4,0001969a913d9ada268d363b54afa383,unknown,50,50.00,1,1,NaN,2019-12-10,2019-12-10,0,0.0


In [71]:
df_grouped['cancelled_percentage'].value_counts()

cancelled_percentage
0.0     141268
50.0     13948
66.7      3373
33.3      2691
75.0      1246
         ...  
93.1         1
45.1         1
43.3         1
56.0         1
68.4         1
Name: count, Length: 293, dtype: int64

In [72]:
# посмотрим наиболее "позднюю" дату в дашем датафрейме
df_grouped['last_transaction_date'].sort_values(ascending=False)

146915   2021-06-30
110764   2021-06-30
76029    2021-06-30
41495    2021-06-30
133586   2021-06-30
            ...    
116914   2012-03-02
64461    2012-03-02
165144   2012-03-01
42353    2012-03-01
74061    2012-02-25
Name: last_transaction_date, Length: 168378, dtype: datetime64[ns]

Мы видим, что данные у нас есть по 30.06.2021. т.е. по конец первого полугодия 2021 года. На сегодняшний день, очевидно, что данные являются устаревшими для реального прогноза оттока или даты следующего платежа по тому или иному донору. Собственно и отток, и следующие платежи уже прошли и остались в прошлом за прошедшие с момента сбора данных три с лишним года. Поэтому для подготовки нашей модели будем считать, что сегодня у нас 01.07.2021 г. и наши данные выгружены вот буквально только что.

Теперь нужно определиться с тем, каких доноров мы будем считать активными на данный момент. Тех, у кого есть хотя бы 2  успешных транзакции, сделанные не в 1 день
 ( а значит есть и интервал между ними, а соответственно, и средний интервал между успешными транзакциями) будем считать активными при условии, что с последней успешной транзакции по сегодняшнее число (01.07.2021) прошло времени не более среднего интервала между транзакциями для этого самого донора. Доноров у которых только 1 успешная транзакция, а значит нет никаких интервалов, будем считать активными, если с момента их единственной транзакции по сегодняшнее число (01.07.2021) прошло времени не более общего среднего значения интервала по всем донорам из нашего датафрейма df_grouped.

In [73]:
# заменяем все NULL значения в колонке 'mean_date_diff' на нули
df_grouped['mean_date_diff'] = df_grouped['mean_date_diff'].fillna(0)

In [74]:
# Вычисляем среднее количество дней между успешными транзакциями в целом по датафрейму.
#  Сумму всех средних интервалов, на количество доноров, чьи интервалы больше 0.
print(f'Среднее значение интервала между успешными транзакциями по всему датафрейму: {round(df_grouped['mean_date_diff'].sum()/ len(df_grouped[df_grouped['mean_date_diff'] > 0]))} дней')

Среднее значение интервала между успешными транзакциями по всему датафрейму: 166 дней


In [75]:
# заменим все нулевые средние интервалы на полученное выше значение
df_grouped['mean_date_diff'] = df_grouped['mean_date_diff'].replace(0, 166).astype(int)

In [76]:
# Введем переменную для текущей даты
comparison_date = pd.to_datetime('2021-07-01')

In [77]:
comparison_date

Timestamp('2021-07-01 00:00:00')

Добавляем колонку 'churn' которая фактически означает что донор не активен (отток) (1- не активен(отток), 0 - активен).  За отток донора принимаем условие, что разница между 01.07.2021 
 и днем последней транзакции донора больше, чем среднее значение интервала между платежами (mean_date_diff) данного донора. Эта колонка будет у нас целевой переменной при прогнозировании оттока с помощью машинного обучения

In [78]:

df_grouped['churn'] = ((comparison_date - df_grouped['last_transaction_date']) > pd.to_timedelta(df_grouped['mean_date_diff'], unit='d')).astype(int)

In [79]:
# Проверяем результат
df_grouped.head()

,donor_id,sex,sum_amount,mean_amount,count_transactions_completed,count_reccurent_transactions,mean_date_diff,first_transaction_date,last_transaction_date,count_transactions_cancelled,cancelled_percentage,churn
0,000085233064aac73f54ed9292e76613,f,6330,2110.00,3,0,610,2017-12-18,2021-04-22,0,0.0,0
1,00008e5479b05b0cf84620b017a54f6e,unknown,200,200.00,1,1,166,2020-05-04,2020-05-04,0,0.0,1
2,0000f19ad7b4da36765dacad76cdadc9,f,800,400.00,2,0,35,2017-10-03,2017-11-07,0,0.0,1
3,0000f6c6b8f0a8c8cdbcaa2330ea7f64,f,8600,296.55,29,28,31,2018-12-23,2021-06-02,0,0.0,0
4,0001969a913d9ada268d363b54afa383,unknown,50,50.00,1,1,166,2019-12-10,2019-12-10,0,0.0,1


Добавим категориальную колонку 'has_reccure', которая означает является ли донор реккурентным подписчиком. 
Для этого будут учитываться только те доноры, которые делали реккуретные платежи и которые являются активными (churn=0)

In [80]:

#df_grouped['has_reccure'] =  ((df_grouped['count_reccurent_transactions']) > 0).astype(int)
df_grouped['has_reccure'] = np.where((df_grouped['churn']==0) & (df_grouped['count_reccurent_transactions'] > 0), 1, 0)

In [81]:
df_grouped.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 168378 entries, 0 to 168377
Data columns (total 13 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   donor_id                      168378 non-null  object        
 1   sex                           168378 non-null  object        
 2   sum_amount                    168378 non-null  int32         
 3   mean_amount                   168378 non-null  float64       
 4   count_transactions_completed  168378 non-null  int64         
 5   count_reccurent_transactions  168378 non-null  int32         
 6   mean_date_diff                168378 non-null  int32         
 7   first_transaction_date        168378 non-null  datetime64[ns]
 8   last_transaction_date         168378 non-null  datetime64[ns]
 9   count_transactions_cancelled  168378 non-null  int32         
 10  cancelled_percentage          168378 non-null  float64       
 11  churn        

In [82]:
# Преобразуем все поля с датами в целые числа  с помощью функции toordinal для использования в качестве дискретного признака в модели машинного обучения
df_grouped['last_transaction_date'] = df_grouped['last_transaction_date'].apply(lambda x: x.date().toordinal())
df_grouped['first_transaction_date'] = df_grouped['first_transaction_date'].apply(lambda x: x.date().toordinal())

In [83]:
# вводим dummy-переменные для категориальных признаков. В нашем датасете это только sex( Пол донора). Donor_id мы в состав признаков включать не будем.
dummies = pd.get_dummies(df_grouped['sex'], prefix='sex')
df_grouped = pd.concat([df_grouped, dummies], axis=1)

In [84]:
# Конвертируем новые признаки из логических в целочисленные
df_grouped['sex_f'] = df_grouped['sex_f'].astype(int)
df_grouped['sex_m'] = df_grouped['sex_m'].astype(int)
df_grouped['sex_unknown'] = df_grouped['sex_unknown'].astype(int)

In [85]:
# Удалим колонку 'sex'. После создания dummy-переменных для нее, она нам больше не нужна

df_grouped.drop('sex', axis=1, inplace=True)

In [86]:
# Cмотрим итоговый результат. Именно в таком виде будем выгружать данные в CSV файл для дальнейшего его подключения в качестве источника в модели AW.
df_grouped

,donor_id,sum_amount,mean_amount,count_transactions_completed,count_reccurent_transactions,mean_date_diff,first_transaction_date,last_transaction_date,count_transactions_cancelled,cancelled_percentage,churn,has_reccure,sex_f,sex_m,sex_unknown
0,000085233064aac73f54ed9292e76613,6330,2110.00,3,0,610,736681,737902,0,0.0,0,0,1,0,0
1,00008e5479b05b0cf84620b017a54f6e,200,200.00,1,1,166,737549,737549,0,0.0,1,0,0,0,1
2,0000f19ad7b4da36765dacad76cdadc9,800,400.00,2,0,35,736605,736640,0,0.0,1,0,1,0,0
3,0000f6c6b8f0a8c8cdbcaa2330ea7f64,8600,296.55,29,28,31,737051,737943,0,0.0,0,1,1,0,0
4,0001969a913d9ada268d363b54afa383,50,50.00,1,1,166,737403,737403,0,0.0,1,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
168373,fffdc2d9ebcd2b03740a6bcfe9461a5b,100,100.00,1,1,166,737955,737955,0,0.0,0,1,1,0,0
168374,fffe4b8e50a941e677ec80535eaba4f9,300,300.00,1,1,166,737488,737488,0,0.0,1,0,0,0,1
168375,fffeb1bf939cf19731a442e92f9a2175,8500,500.00,17,16,38,737048,737657,1,5.6,1,0,1,0,0
168376,ffff964599367cf1498f1e19aa3b0337,500,500.00,1,0,166,737959,737959,0,0.0,0,0,1,0,0


**Выгружаем получившийся  сгруппированный датасет в CSV-файл**

In [87]:
# очищенные, обогащенные, сгруппированные данные для загрузки в модель AW
df_grouped.to_csv('df_donor_trans_for_ml.csv', index=False)